update user settings - issue 20

JoeArmani opened the pull request:
Link to issue: [https://github.com/episphere/connectApp/issues/20](url)

This PR addresses connectApp issue #20 - allow members to update their profile information once they’re verified.

**Functionality:**

Profile must be verified to be editable.

- If submitted, user can see their data. If verified, user can edit their data.

Refactored HTML for readability, separation of concerns, and faster dev access

Integrated conceptId mapping for readability.

Created toggling edit & submit buttons for name, contact info, and mailing address.

Updated Name display to include additional fields based on user’s data availability.

Created Name Form.

- 	dynamically load fields based on user’s information.
- 	validate client-side.
- 	integrated data submit.

Updated Contact Information display to include additional fields based on user’s data.

Created Contact information form.

- 	dynamically load fields based on user’s information.
- 	validate client-side.
- 	integrated data submit.

Created Mailing address update form.

- 	uses Google Maps API for easy and correct address data.
- 	validates client-side.
- 	integrates data submit.

Updated Sign-In Information to match code stylistically, so settings.js and settingsHelper.js are cohesive moving forward.

Notes based on original issue:

- Salutation conceptId not found
- WhatsApp conceptId not found

we-ai commented on js/pages/settings.js:
```diff
@@ -1,327 +1,996 @@
+  additionalEmail = userData[cId.additionalEmail];
+  let template = '';
+
+  if (myData.code === 200 && userData[cId.userProfileSubmittedAutogen] !== undefined && userData[cId.userProfileSubmittedAutogen] === cId.yes) {
```
```suggestion
  if (myData.code === 200 && userData[cId.userProfileSubmittedAutogen] === cId.yes) {
```

we-ai commented on js/pages/settings.js:
```diff
@@ -1,327 +1,996 @@
+   */
+  if (
+    myData.code === 200 &&
+    userData[cId.userProfileSubmittedAutogen] !== undefined &&
+    userData[cId.userProfileSubmittedAutogen] === cId.yes &&
+    userData[cId.verified] !== undefined &&
+    userData[cId.verified] === cId.yes
```
```suggestion
    userData[cId.userProfileSubmittedAutogen] === cId.yes &&
    userData[cId.verified] === cId.yes
```

we-ai commented on js/settingsHelpers.js:
```diff
@@ -0,0 +1,408 @@
+  EMAIL: 'emailForm',
+};
+
+export const validEmailFormat = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
```
We have a `validEmailFormat` in `shared.js` file.

we-ai replied:
`validEmailFormat` in `shared.js` is not enforcing top level domains. I can update it. There's a related spec about email format [here](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address)

JoeArmani replied:
resolved with regexp in `shared.js` that enforces TLD (min 2-char). It validates Unicode and Punycode-encoded Unicode.

jhflorey commented on js/pages/settings.js:
```diff
@@ -1,327 +1,996 @@
+  document.title = 'My Connect - My Profile';
+  showAnimation();
+  const myData = await getMyData();
+  const userData = myData.data;
```
Should we check the data exists

JoeArmani replied:
On one hand the user doc is written by cloud functions and the user is authenticated at this point in the app flow. Chances are almost nonexistent that there won't be a user doc. On the other hand chances are probably not zero - it seems like a worthwhile check - I agree with you and I'll add this. Thanks 👍

jhflorey commented on js/settingsHelpers.js:
```diff
@@ -0,0 +1,408 @@
+
+export const validEmailFormat = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
+
+export const updatePhoneNumberInputFocus = () => {
```
Should we create common function like

const focusEvent = (id) => {
const element = document.getElementById(id);
  element.addEventListener('keyup', () => {
    if (element.value.trim().length === 3) {
      element.nextElementSibling.nextElementSibling.focus();
    }
  });
}

JoeArmani replied:
Great suggestion. It's only called on update contact information, so I'll just iterate the array.

`
export const updatePhoneNumberInputFocus = () => {
  const phoneElementIds = ['mobilePhoneNumber1', 'mobilePhoneNumber2', 'homePhoneNumber1', 'homePhoneNumber2', 'otherPhoneNumber1', 'otherPhoneNumber2'];

  const initFocusHandler = elementId => {
    const element = document.getElementById(elementId);
    element.addEventListener('keyup', () => {
      if (element.value.trim().length === 3) {
        element.nextElementSibling.nextElementSibling.focus();
      }
    });
  };

  phoneElementIds.forEach(initFocusHandler);
};
`

JoeArmani commented:
Updates in 2nd commit:
-resolved all comments/suggestions to date
-updated `getMyData()` in `shared.js` to include error handling.
-move `handler functions` outside of `renderSettingsPage()` in `settings.js`
-added check if data exists after `getMyData()` call. 
-updated variable scoping and page loading logic to accommodate above changes.

JoeArmani commented:
Updates in 3rd commit:
-resolved phone number flexibility to mimic sign up flow: user can enter mobile, home, or other phone number. Only one is required. User has the option to enter two or all three, now the UX plays nicely with those options.
-expanded upon name validation to include some Latin characters and special characters.
-updated edit form styling for readability
-added `canWeVoicemail` field for `otherPhoneNumber`

JoeArmani commented:
Final commit:
-enforce whatwg.org email validation standard + TLD requirement.
-clean up style.css
